appstudio-controller/controllers/appstudio.redhat.com/applicationpromotionrun_controller.go

```diff
@@ -195,29 +292,54 @@ func (r *ApplicationPromotionRunReconciler) ReconcilePOC(ctx context.Context, re
 		if !match {
 			waitingGitOpsDeployments = append(waitingGitOpsDeployments, gitopsDeployment.Name)
 			continue
-		}
+		}*/
+	}
+
+	// Check time limit set for PromotionRun reconcilation, fail if the conditions aren't met in the given time frame.
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(promotionRun), promotionRun); err != nil {
+		log.Error(err, "unable to retrieve promotionRun: "+promotionRun.Name)
+		return ctrl.Result{}, fmt.Errorf("unable to retrieve promotionRun '%s', %v", promotionRun.Name, err)
 	}
 
-	// TODO: GITOPSRVCE-157 - Implement a time limit on the spec of Promotion Run, and fail if the conditions aren't met in the timeframe.
+	if !promotionRun.Status.PromotionStartTime.IsZero() && (metav1.Now().Sub(promotionRun.Status.PromotionStartTime.Time).Minutes() > PromotionRunTimeOutLimit) {
+
+		promotionRun.Status.CompletionResult = appstudioshared.PromotionRunCompleteResult_Failure
+		promotionRun.Status.State = appstudioshared.PromotionRunState_Complete
+
+		// Update Status.Environment.Status field.
+		if err = updateStatusEnvironmentStatus(ctx, r.Client, fmt.Sprintf("Promotion Failed. Could not be completed in %d Minutes.", PromotionRunTimeOutLimit),
+			promotionRun, appstudioshared.ApplicationPromotionRunEnvironmentStatus_Failed, log); err != nil {
+			log.Error(err, "unable to update PromotionRun environment status: "+promotionRun.Name)
+			return ctrl.Result{}, fmt.Errorf("unable to update promotionRun %v", err)
+		}
+		return ctrl.Result{}, nil
+	}
 
 	if len(waitingGitOpsDeployments) > 0 {
-		fmt.Println("Waiting for GitOpsDeployments to have expected commit/sync/health:", waitingGitOpsDeployments)
+		log.Info("Waiting for GitOpsDeployments to have expected commit/sync/health:" + strings.Join(waitingGitOpsDeployments[:], ", "))
+
+		promotionRun.Status.State = appstudioshared.PromotionRunState_Waiting
+
+		// Update Status.Environment.Status field.
+		if err = updateStatusEnvironmentStatus(ctx, r.Client, "Waiting for following GitOpsDeployments to be Synced/Healthy: "+strings.Join(waitingGitOpsDeployments[:], ", "),
+			promotionRun, appstudioshared.ApplicationPromotionRunEnvironmentStatus_InProgress, log); err != nil {
+			log.Error(err, "unable to update PromotionRun environment status: "+promotionRun.Name)
+			return ctrl.Result{}, fmt.Errorf("unable to update promotionRun %v", err)
+		}
 		return ctrl.Result{RequeueAfter: time.Second * 15}, nil
 	}
 
 	// All the GitOpsDeployments are synced/healthy, and they are synced with a commit that includes the target snapshot.
```

Comment:
If the Reconcile() func is successful (it didn't hit any errors), then we should make sure that the PromotionRunConditionErrorOccurred condition is reset to false (PromotionRunConditionStatusFalse).
- For example, it looks like at the moment  aren't any references to `PromotionRunConditionStatusFalse` in the code, except where it is defined), so when it is set to True it will never be reset back to False (unless I am missing logic for this).

With this behaviour, if an error DID occur, but has now been fixed, the `ErrorOccurred` condtion will be reset back to `False` (indicating no error).

For example, we can reset the condition in these spots:
```go
	if len(waitingGitOpsDeployments) > 0 {
		log.Info("Waiting for GitOpsDeployments to have expected commit/sync/health:" + strings.Join(waitingGitOpsDeployments[:], ", "))
		// (...)
		// set ErrorOccurred condition to false:
		updateStatusConditions(ctx, r.Client, "", promotionRun, appstudioshared.PromotionRunConditionErrorOccurred, appstudioshared.PromotionRunConditionStatusFalse, "")
	}
```
If we are waiting for GitOpsDeployments, then no error has occurred, so we can set the status to false.

Likewise, once the GitOpsDeployments are synced/healthy, we can se the condition to false:
```go
// All the GitOpsDeployments are synced/healthy, and they are synced with a commit that includes the target snapshot.
// set ErrorOccurred condition to false
updateStatusConditions(ctx, r.Client, "", promotionRun, appstudioshared.PromotionRunConditionErrorOccurred, appstudioshared.PromotionRunConditionStatusFalse, "")
```

It likewise couldn't hurt to have a unit test to verify that the condition goes from True, then back to False, once the issue is fixed (my guess is it can be added to one of our existing unit tests).